README.md
# bip32

A [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) compatible library for Flutter writing in Dart.

Inspired by [bitcoinjs](https://github.com/bitcoinjs/bip32)

## Example
```dart

import 'package:dart_bip32/dart_bip32.dart' as bip32;
import 'package:hex/hex.dart';

main() {
  bip32.BIP32 node = bip32.BIP32.fromBase58('xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi');

  print(HEX.encode(node.privateKey));
  // => e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35

  bip32.BIP32 nodeNeutered = node.neutered();
  print(nodeNeutered.isNeutered());
  // => true

  print(HEX.encode(nodeNeutered.publicKey));
  // => 0339a36013301597daef41fbe593a02cc513d0b55527ec2df1050e2e8ff49c85c2

  print(nodeNeutered.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  bip32.BIP32 child = node.derivePath('m/0/0');
  print(child.toBase58());
  // => xprv9ww7sMFLzJMzur2oEQDB642fbsMS4q6JRraMVTrM9bTWBq7NDS8ZpmsKVB4YF3mZecqax1fjnsPF19xnsJNfRp4RSyexacULXMKowSACTRc

  print(HEX.encode(child.privateKey));
  // => f26cf12f89ab91aeeb8d7324a22e8ba080829db15c9245414b073a8c342322aa

  bip32.BIP32 childNeutered = child.neutered();
  print(childNeutered.isNeutered());
  // => true

  print(HEX.encode(childNeutered.publicKey));
  // => 02756de182c5dd4b717ea87e693006da62dbb3cddaa4a5cad2ed1f5bbab755f0f5

  print(childNeutered.toBase58());
  // => xpub6AvUGrnEpfvJ8L7GLRkBTByQ9uBvUHp9o5VxHrFxhvzV4dSWkySpNaBoLR9FpbnwRmTa69yLHF3QfcaxbWT7gWdwws5k4dpmJvqpEuMWwnj

  bip32.BIP32 nodeFromSeed = bip32.BIP32.fromSeed(HEX.decode("000102030405060708090a0b0c0d0e0f"));
  print(nodeFromSeed.toBase58());
  // => xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi

  bip32.BIP32 nodeFromPub = bip32.BIP32.fromBase58("xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8");
  print(nodeFromPub.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  var message = HEX.decode("0202020202020202020202020202020202020202020202020202020202020202");
  var signature = nodeFromSeed.sign(message);
  print(signature);
  // => [63, 219, 20, 114, 95, 184, 192, 55, 216, 206, 126, 121, 17, 71, 64, 70, 163, 82, 247, 73, 243, 95, 30, 137, 177, 155, 100, 225, 177, 203, 217, 147, 122, 64, 208, 129, 54, 133, 113, 41, 216, 160, 191, 15, 136, 98, 235, 25, 219, 178, 70, 222, 127, 151, 135, 242, 25, 192, 161, 187, 187, 84, 81, 215]

  print(HEX.encode(signature));
  // => 3fdb14725fb8c037d8ce7e7911474046a352f749f35f1e89b19b64e1b1cbd9937a40d08136857129d8a0bf0f8862eb19dbb246de7f9787f219c0a1bbbb5451d7

  print(nodeFromSeed.verify(message, signature));
  // => true
}

```

## Contributor
  * anicdh <anic.dh@gmail.com>
  * p3root 
  * jcramer (null-safety)


CHANGELOG.md
## 1.0.0

- Initial version, created by anicdh

## 1.0.1

- Format code & add description

## 1.0.2

- Change version bs58check

## 1.0.3

- Change generate BIP32 from function to factory

## 1.0.4

- Fix error message when generate BIP32 from private key
- Fix argument of wif decode function

## 1.0.5

- Fix wif decodeRaw function

## 1.0.6

- Fix derive bug

## 1.0.8

- Update pointycastle version to 2.0.0

## 1.0.10

- Fix sign

## 2.0.0
- Add null-safety

## 2.0.1
- update dart_bs58check package

## 2.0.2
- update dart_wif package

## 2.0.3
- update example file name

## 2.0.4
- update dart_bs58check package version
- 
## 2.0.5
- export utils

test/ecc_test.dart
import 'dart:typed_data';

import 'package:dart_bip32/dart_bip32.dart';
import 'package:dart_bip32/src/utils/ecurve.dart' as ecc;
import 'package:hex/hex.dart';
import 'package:test/test.dart';

final defiChainTestnet = NetworkType(
    bip32: Bip32Type(private: 0x04358394, public: 0x043587cf), wif: 0xef);

void main() {
  group("curve test", () {
    test("test private key generation", () {
      // seed
      //55b18e96ce3964ef2c81ad69249eca6d42682c11fbe525df6671fcbf0c2be902 private key

      final hdSeed = BIP32.fromSeed(
          HEX.decode(
                  "6607599b768ce88470b3b20919f9c63bff663e2f1ec3e3072d22fd9da3847784c361d5accc3b411019f5c81dd3e4ccf9fd1fddb232bfc9bfe23864e2e6ee793f")
              as Uint8List,
          defiChainTestnet);

      final xMasterPriv = BIP32.fromSeed(hdSeed.privateKey!, defiChainTestnet);
      final privateKey = xMasterPriv.derivePath("m/0'/0'/0'");

      final privateKeyHex = HEX.encode(privateKey.privateKey!);

      expect("55b18e96ce3964ef2c81ad69249eca6d42682c11fbe525df6671fcbf0c2be902",
          privateKeyHex);
    });

    test("test sign", () {
      ecc.sign(
          HEX.decode(
                  "b11d3d5e4ae12b89d5e3872ccc7d1f96d29b0ab888b67dccf1be5164b811cdbe")
              as Uint8List,
          HEX.decode(
                  "55b18e96ce3964ef2c81ad69249eca6d42682c11fbe525df6671fcbf0c2be902")
              as Uint8List);
    });
  });

  group('Failing ecc test', () {
    test('Should fail with the third set of key-msg', () {
      String key, msg;
      Uint8List sig;

      key = 'f92ca9fe5f77afa489214a7ba2bd6b36d30dd4acdb55c70d1378b5c45c831820';
      msg = '045a7448dffff67c08023d16279c57c0bd16af6467580c183cc4672e768b8a77';

      sig =
          ecc.sign(HEX.decode(msg) as Uint8List, HEX.decode(key) as Uint8List);
      print(HEX.encode(sig));

      key = '8c39fb3d889b6be22850254dc7ce3247c559d9f968785d88251ee4457633e335';
      msg = '6ad99bc30926fd79cf5a977044088fd298a59728c5b0c1040713d3a8e1c1e69b';

      sig =
          ecc.sign(HEX.decode(msg) as Uint8List, HEX.decode(key) as Uint8List);
      print(HEX.encode(sig));

      key = '2de619ea940d31927e0a7bbd5ebf855ab66f488b94a65cf929a1eb70b54ec771';
      msg = 'd82a6db2583353856dd4dadc38cdb6373b092ea1f026fa04f03aa3c4e972454f';

      sig =
          ecc.sign(HEX.decode(msg) as Uint8List, HEX.decode(key) as Uint8List);
      print(HEX.encode(sig));
    });
  });
}


test/bip32_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:dart_bip32/dart_bip32.dart';
import 'package:hex/hex.dart';
import 'package:test/test.dart';

final litecoin = NetworkType(
    bip32: Bip32Type(private: 0x019d9cfe, public: 0x019da462), wif: 0xb0);
List<dynamic> validAll = [];

void main() {
  Map<String, dynamic> fixtures = json
      .decode(File('./test/fixtures.json').readAsStringSync(encoding: utf8));
  for (var f in (fixtures['valid'] as List<dynamic>)) {
    f['master']['network'] = f['network'];
    f['master']['children'] = f['children'];
    f['master']['comment'] = f['comment'];
    for (var fc in (f['children'] as List<dynamic>)) {
      fc['network'] = f['network'];
      validAll.add(fc);
    }
    validAll.add(f['master']);
    for (var ff in validAll) {
      group(ff['comment'] ?? ff['base58Priv'], () {
        setUp(() {});
        NetworkType? network;
        if (ff['network'] == 'litecoin') {
          network = litecoin;
        }
        var hdPrv = BIP32.fromBase58(ff['base58Priv'], network);
        test('works for private key -> HD wallet', () {
          verify(hdPrv, true, ff, network);
        });

        var hdPub = BIP32.fromBase58(ff['base58'], network);
        test('works for public key -> HD wallet', () {
          verify(hdPub, false, ff, network);
        });

        if (ff['seed'] != null) {
          var seed = HEX.decode(ff['seed']);
          var hdSeed = BIP32.fromSeed(seed as Uint8List, network);
          test('works for seed -> HD wallet', () {
            verify(hdSeed, true, ff, network);
          });
        }
      });
    }
  }

  test('fromBase58 throws', () {
    for (var f in (fixtures['invalid']['fromBase58'] as List<dynamic>)) {
      NetworkType? network;
      if (f['network'] != null && f['network'] == 'litecoin') {
        network = litecoin;
      }
      BIP32? hd;
      try {
        hd = BIP32.fromBase58(f['string'], network);
      } catch (err) {
        expect((err as ArgumentError).message, f['exception']);
      } finally {
        expect(hd, null);
      }
    }
  });

  test('works for Private -> public (neutered)', () {
    final f = fixtures['valid'][1];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58Priv'] as String);
    final child = master.derive(c['m']).neutered();
    expect(child.toBase58(), c['base58']);
  });

  test('works for Private -> public (neutered, hardened)', () {
    final f = fixtures['valid'][0];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58Priv'] as String);
    final child = master.deriveHardened(c['m']).neutered();
    expect(child.toBase58(), c['base58']);
  });

  test('works for Public -> public', () {
    final f = fixtures['valid'][1];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58'] as String);
    final child = master.derive(c['m']);
    expect(child.toBase58(), c['base58']);
  });

  test('throws on Public -> public (hardened)', () {
    final f = fixtures['valid'][0];
    final c = f['master']['children'][0];
    final master = BIP32.fromBase58(f['master']['base58'] as String);
    BIP32? hd;
    try {
      hd = master.deriveHardened(c['m']);
    } catch (err) {
      expect((err as ArgumentError).message,
          "Missing private key for hardened child key");
    } finally {
      expect(hd, null);
    }
  });

  test('throws on wrong types', () {
    final f = fixtures['valid'][0];
    final master = BIP32.fromBase58(f['master']['base58'] as String);
    for (var fx in (fixtures['invalid']['derive'] as List<dynamic>)) {
      BIP32? hd;
      try {
        hd = master.derive(fx);
      } catch (err) {
        expect((err as ArgumentError).message, "Expected UInt32");
      } finally {
        expect(hd, null);
      }
    }
    for (var fx in (fixtures['invalid']['deriveHardened'] as List<dynamic>)) {
      BIP32? hd;
      try {
        hd = master.deriveHardened(fx);
      } catch (err) {
        expect((err as ArgumentError).message, "Expected UInt31");
      } finally {
        expect(hd, null);
      }
    }
    for (var fx in (fixtures['invalid']['derivePath'] as List<dynamic>)) {
      BIP32? hd;
      try {
        hd = master.derivePath(fx);
      } catch (err) {
        expect((err as ArgumentError).message, "Expected BIP32 Path");
      } finally {
        expect(hd, null);
      }
    }
    BIP32? hdFPrv1, hdFPrv2;
    final zero32 = Uint8List.fromList(List.generate(32, (index) => 0));
    final one32 = Uint8List.fromList(List.generate(32, (index) => 1));
    try {
      hdFPrv1 = BIP32.fromPrivateKey(Uint8List(2), one32);
    } catch (err) {
      expect((err as ArgumentError).message,
          "Expected property privateKey of type Buffer(Length: 32)");
    } finally {
      expect(hdFPrv1, null);
    }
    try {
      hdFPrv2 = BIP32.fromPrivateKey(zero32, one32);
    } catch (err) {
      expect((err as ArgumentError).message, "Private key not in range [1, n]");
    } finally {
      expect(hdFPrv2, null);
    }
  });

  test("works when private key has leading zeros", () {
    const key =
        "xprv9s21ZrQH143K3ckY9DgU79uMTJkQRLdbCCVDh81SnxTgPzLLGax6uHeBULTtaEtcAvKjXfT7ZWtHzKjTpujMkUd9dDb8msDeAfnJxrgAYhr";
    BIP32 hdkey = BIP32.fromBase58(key);
    expect(HEX.encode(hdkey.privateKey!),
        "00000055378cf5fafb56c711c674143f9b0ee82ab0ba2924f19b64f5ae7cdbfd");
    BIP32 child = hdkey.derivePath("m/44'/0'/0'/0/0'");
    expect(HEX.encode(child.privateKey!),
        "3348069561d2a0fb925e74bf198762acc47dce7db27372257d2d959a9e6f8aeb");
  });

  test('derive', () {
    final hd = BIP32.fromBase58(
        'xprv9s21ZrQH143K3Jpuz63XbuGs9CH9xG4sniVBBRVm6AJR57D9arxWz6FkXF3JSxSK7jUmVA11AdWa6ZsUtwGztE4QT5i8Y457RRPvMCc39rY');
    final d = hd.derivePath("m/1'/199007533'/627785449'/1521366139'/1'");
    expect(d.toBase58(),
        'xprvA39a1i4ieYqGUQ7G1KGnaGzGwm7v3emjms3QN4jZ3HPeubXjshA3XjD5XFaiNgWFvoyC2NV5jN4eFcsVhkrWkvwR4qjdPbue3kpt6Ur3JRf');
  });

  test("fromSeed", () {
    for (var f in (fixtures['invalid']['fromSeed'] as List<dynamic>)) {
      BIP32? hd;
      try {
        hd = BIP32.fromSeed(HEX.decode(f['seed']) as Uint8List);
      } catch (err) {
        expect((err as ArgumentError).message, f['exception']);
      } finally {
        expect(hd, null);
      }
    }
  });

  test("ecdsa", () {
    Uint8List seed = Uint8List.fromList(List.generate(32, (index) => 1));
    Uint8List hash = Uint8List.fromList(List.generate(32, (index) => 2));
    String sigStr =
        "9636ee2fac31b795a308856b821ebe297dda7b28220fb46ea1fbbd7285977cc04c82b734956246a0f15a9698f03f546d8d96fe006c8e7bd2256ca7c8229e6f5c";
    Uint8List signature = HEX.decode(sigStr) as Uint8List;
    BIP32 node = BIP32.fromSeed(seed);
    expect(HEX.encode(node.sign(hash)), sigStr);
    expect(node.verify(hash, signature), true);
    expect(node.verify(seed, signature), false);
  });
}

void verify(BIP32 hd, prv, f, network) {
  expect(HEX.encode(hd.chainCode), f['chainCode']);
  expect(hd.depth, f['depth'] ?? 0);
  expect(hd.index, f['index'] ?? 0);
  expect(HEX.encode(hd.fingerprint), f['fingerprint']);
  expect(HEX.encode(hd.identifier), f['identifier']);
  expect(HEX.encode(hd.publicKey), f['pubKey']);
  if (prv) {
    expect(hd.toBase58(), f['base58Priv']);
    expect(HEX.encode(hd.privateKey!), f['privKey']);
    expect(hd.toWIF(), f['wif']);
  } else {
    expect(hd.privateKey, null);
  }
  expect(hd.neutered().toBase58(), f['base58']);
  expect(hd.isNeutered(), !prv);

  if (f['children'] == null) return;
  if (!prv &&
      (f['children'] as List<dynamic>)
          .map((fc) => fc['hardened'])
          .contains(true)) return;

  for (var cf in (f['children'] as List<dynamic>)) {
    var chd = hd.derivePath(cf['path']);
    verify(chd, prv, cf, network);
    var chdNoM = hd.derivePath((cf['path'] as String).substring(2)); // no m/
    verify(chdNoM, prv, cf, network);
  }

  // test deriving path from successive children
  var shd = hd;
  for (var cf in (f['children'] as List<dynamic>)) {
    if (cf['m'] == null) return;
    if (cf['hardened'] != null && cf['hardened'] as bool) {
      shd = shd.deriveHardened(cf['m']);
    } else {
      // verify any publicly derived children
      if (cf['base58'] != null) {
        verify(shd.neutered().derive(cf['m']), false, cf, network);
      }
      shd = shd.derive(cf['m']);
      verify(shd, prv, cf, network);
    }
  }
}


lib/src/utils/ecurve.dart
import 'dart:typed_data';

import 'package:hex/hex.dart';
import "package:pointycastle/api.dart"
    show PrivateKeyParameter, PublicKeyParameter;
import "package:pointycastle/digests/sha256.dart";
import 'package:pointycastle/ecc/api.dart'
    show ECPrivateKey, ECPublicKey, ECSignature, ECPoint;
import "package:pointycastle/ecc/curves/secp256k1.dart";
import 'package:pointycastle/macs/hmac.dart';
import "package:pointycastle/signers/ecdsa_signer.dart";

final _negativeFlag = BigInt.from(0x80);

final _zero32 = Uint8List.fromList(List.generate(32, (index) => 0));
final _ecGroupOrder = HEX
    .decode("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141");
final _ecpP = HEX
    .decode("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f");
final _secp256k1 = ECCurve_secp256k1();
final _n = _secp256k1.n;
final _g = _secp256k1.G;
BigInt _nDiv2 = _n >> 1;

// Exceptions
/// Bad Private Key
const _throwBadPrivate = 'Expected Private';

/// Bad Point
const _throwBadPoint = 'Expected Point';

/// Bad Tweak
const _throwBadTweak = 'Expected Tweak';

/// Bad Hash
const _throwBadHash = 'Expected Hash';

/// Bad Signature
const _throwBadSignature = 'Expected Signature';

/// Is Buffer Private Key ?
bool isPrivate(Uint8List x) {
  if (!isScalar(x)) return false;
  return _compare(x, _zero32) > 0 && // > 0
      _compare(x, _ecGroupOrder as Uint8List) < 0; // < G
}

/// Is Buffer Point
bool isPoint(Uint8List p) {
  if (p.length < 33) {
    return false;
  }
  var t = p[0];
  var x = p.sublist(1, 33);

  if (_compare(x, _zero32) == 0) {
    return false;
  }
  if (_compare(x, _ecpP as Uint8List) == 1) {
    return false;
  }
  try {
    decodeFrom(p);
  } catch (err) {
    return false;
  }
  if ((t == 0x02 || t == 0x03) && p.length == 33) {
    return true;
  }
  var y = p.sublist(33);
  if (_compare(y, _zero32) == 0) {
    return false;
  }
  if (_compare(y, _ecpP as Uint8List) == 1) {
    return false;
  }
  if (t == 0x04 && p.length == 65) {
    return true;
  }
  return false;
}

/// Is Scalar
bool isScalar(Uint8List x) {
  return x.length == 32;
}

/// Is Order Scalar
bool isOrderScalar(x) {
  if (!isScalar(x)) return false;
  return _compare(x, _ecGroupOrder as Uint8List) < 0; // < G
}

/// Is Signature
bool isSignature(Uint8List value) {
  Uint8List r = value.sublist(0, 32);
  Uint8List s = value.sublist(32, 64);

  return value.length == 64 &&
      _compare(r, _ecGroupOrder as Uint8List) < 0 &&
      _compare(s, _ecGroupOrder as Uint8List) < 0;
}

/// Is Point Compressed
bool _isPointCompressed(Uint8List p) {
  return p[0] != 0x04;
}

/// Assume Compression
bool assumeCompression(bool? value, Uint8List? pubkey) {
  if (value == null && pubkey != null) return _isPointCompressed(pubkey);
  if (value == null) return true;
  return value;
}

/// Point From Scalar
Uint8List? pointFromScalar(Uint8List d, bool _compressed) {
  if (!isPrivate(d)) throw ArgumentError(_throwBadPrivate);
  BigInt dd = fromBuffer(d);
  ECPoint pp = (_g * dd) as ECPoint;
  if (pp.isInfinity) return null;
  return getEncoded(pp, _compressed);
}

/// Point Add Scalar
Uint8List? pointAddScalar(Uint8List p, Uint8List tweak, bool _compressed) {
  if (!isPoint(p)) throw ArgumentError(_throwBadPoint);
  if (!isOrderScalar(tweak)) throw ArgumentError(_throwBadTweak);
  bool compressed = assumeCompression(_compressed, p);
  ECPoint? pp = decodeFrom(p);
  if (_compare(tweak, _zero32) == 0) return getEncoded(pp, compressed);
  BigInt tt = fromBuffer(tweak);
  ECPoint qq = (_g * tt) as ECPoint;
  ECPoint uu = (pp! + qq) as ECPoint;
  if (uu.isInfinity) return null;
  return getEncoded(uu, compressed);
}

/// Private Add
Uint8List? privateAdd(Uint8List d, Uint8List tweak) {
  if (!isPrivate(d)) throw ArgumentError(_throwBadPrivate);
  if (!isOrderScalar(tweak)) throw ArgumentError(_throwBadTweak);
  BigInt dd = fromBuffer(d);
  BigInt tt = fromBuffer(tweak);
  Uint8List dt = toBuffer((dd + tt) % _n);

  if (dt.length < 32) {
    Uint8List padLeadingZero = Uint8List(32 - dt.length);
    dt = Uint8List.fromList(padLeadingZero + dt);
  }

  if (!isPrivate(dt)) return null;
  return dt;
}

/// Sign Hash
Uint8List sign(Uint8List hash, Uint8List x) {
  if (!isScalar(hash)) throw ArgumentError(_throwBadHash);
  if (!isPrivate(x)) throw ArgumentError(_throwBadPrivate);
  ECSignature sig = deterministicGenerateK(hash, x);
  Uint8List buffer = Uint8List(64);
  buffer.setRange(0, 32, _encodeBigInt(sig.r));
  BigInt s;
  if (sig.s.compareTo(_nDiv2) > 0) {
    s = _n - sig.s;
  } else {
    s = sig.s;
  }
  buffer.setRange(32, 64, _encodeBigInt(s));
  return buffer;
}

/// Verify Signature
bool verify(Uint8List hash, Uint8List q, Uint8List signature) {
  if (!isScalar(hash)) throw ArgumentError(_throwBadHash);
  if (!isPoint(q)) throw ArgumentError(_throwBadPoint);
  // 1.4.1 Enforce r and s are both integers in the interval [1, n − 1] (1, isSignature enforces '< n - 1')
  if (!isSignature(signature)) throw ArgumentError(_throwBadSignature);

  ECPoint? Q = decodeFrom(q);
  BigInt r = fromBuffer(signature.sublist(0, 32));
  BigInt s = fromBuffer(signature.sublist(32, 64));

  final signer = ECDSASigner(null, HMac(SHA256Digest(), 64));
  signer.init(false, PublicKeyParameter(ECPublicKey(Q, _secp256k1)));
  return signer.verifySignature(hash, ECSignature(r, s));
  /* STEP BY STEP
  // 1.4.1 Enforce r and s are both integers in the interval [1, n − 1] (2, enforces '> 0')
  if (r.compareTo(n) >= 0) return false;
  if (s.compareTo(n) >= 0) return false;

  // 1.4.2 H = Hash(M), already done by the user
  // 1.4.3 e = H
  BigInt e = fromBuffer(hash);

  BigInt sInv = s.modInverse(n);
  BigInt u1 = (e * sInv) % n;
  BigInt u2 = (r * sInv) % n;

  // 1.4.5 Compute R = (xR, yR)
  //               R = u1G + u2Q
  ECPoint R = G * u1 + Q * u2;

  // 1.4.5 (cont.) Enforce R is not at infinity
  if (R.isInfinity) return false;

  // 1.4.6 Convert the field element R.x to an integer
  BigInt xR = R.x.toBigInteger();

  // 1.4.7 Set v = xR mod n
  BigInt v = xR % n;

  // 1.4.8 If v = r, output "valid", and if v != r, output "invalid"
  return v.compareTo(r) == 0;
  */
}

/// Decode a BigInt from bytes in big-endian encoding.
BigInt _decodeBigInt(List<int> bytes) {
  BigInt result = BigInt.from(0);
  for (int i = 0; i < bytes.length; i++) {
    result += BigInt.from(bytes[bytes.length - i - 1]) << (8 * i);
  }
  return result;
}

var _byteMask = BigInt.from(0xff);

/// Encode a BigInt into bytes using big-endian encoding.
Uint8List _encodeBigInt(BigInt number) {
  int needsPaddingByte;
  int rawSize;

  if (number > BigInt.zero) {
    rawSize = (number.bitLength + 7) >> 3;
    needsPaddingByte =
        ((number >> (rawSize - 1) * 8) & _negativeFlag) == _negativeFlag
            ? 1
            : 0;

    if (rawSize < 32) {
      needsPaddingByte = 1;
    }
  } else {
    needsPaddingByte = 0;
    rawSize = (number.bitLength + 8) >> 3;
  }

  final size = rawSize < 32 ? rawSize + needsPaddingByte : rawSize;
  var result = Uint8List(size);
  for (int i = 0; i < size; i++) {
    result[size - i - 1] = (number & _byteMask).toInt();
    number = number >> 8;
  }
  return result;
}

/// [BigInt] from [Uint8List] buffer
BigInt fromBuffer(Uint8List d) {
  return _decodeBigInt(d);
}

/// [BigInt] to [Uint8List] buffer
Uint8List toBuffer(BigInt d) {
  return _encodeBigInt(d);
}

/// [Uint8List] decode to [ECPoint]
ECPoint? decodeFrom(Uint8List P) {
  return _secp256k1.curve.decodePoint(P);
}

/// [ECPoint] encode to [Uint8List]
Uint8List getEncoded(ECPoint? P, [bool compressed = true]) {
  return P!.getEncoded(compressed);
}

/// Deterministic Generate K
ECSignature deterministicGenerateK(Uint8List hash, Uint8List x) {
  final signer = ECDSASigner(null, HMac(SHA256Digest(), 64));
  var pkp = PrivateKeyParameter(ECPrivateKey(_decodeBigInt(x), _secp256k1));
  signer.init(true, pkp);
//  signer.init(false, PublicKeyParameter(ECPublicKey(secp256k1.curve.decodePoint(x), secp256k1)));
  return signer.generateSignature(hash) as ECSignature;
}

/// Compare Two [Uint8List] Buffer
int _compare(Uint8List a, Uint8List b) {
  BigInt aa = fromBuffer(a);
  BigInt bb = fromBuffer(b);
  if (aa == bb) return 0;
  if (aa > bb) return 1;
  return -1;
}


example/dart_bip32_example.dart
import 'dart:typed_data';

import 'package:dart_bip32/dart_bip32.dart' as bip32;
import 'package:hex/hex.dart';

main() {
  bip32.BIP32 node = bip32.BIP32.fromBase58(
      'xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi');

  print(HEX.encode(node.privateKey!));
  // => e8f32e723decf4051aefac8e2c93c9c5b214313817cdb01a1494b917c8436b35

  bip32.BIP32 nodeNeutered = node.neutered();
  print(nodeNeutered.isNeutered());
  // => true

  print(HEX.encode(nodeNeutered.publicKey));
  // => 0339a36013301597daef41fbe593a02cc513d0b55527ec2df1050e2e8ff49c85c2

  print(nodeNeutered.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  bip32.BIP32 child = node.derivePath('m/0/0');
  print(child.toBase58());
  // => xprv9ww7sMFLzJMzur2oEQDB642fbsMS4q6JRraMVTrM9bTWBq7NDS8ZpmsKVB4YF3mZecqax1fjnsPF19xnsJNfRp4RSyexacULXMKowSACTRc

  print(HEX.encode(child.privateKey!));
  // => f26cf12f89ab91aeeb8d7324a22e8ba080829db15c9245414b073a8c342322aa

  bip32.BIP32 childNeutered = child.neutered();
  print(childNeutered.isNeutered());
  // => true

  print(HEX.encode(childNeutered.publicKey));
  // => 02756de182c5dd4b717ea87e693006da62dbb3cddaa4a5cad2ed1f5bbab755f0f5

  print(childNeutered.toBase58());
  // => xpub6AvUGrnEpfvJ8L7GLRkBTByQ9uBvUHp9o5VxHrFxhvzV4dSWkySpNaBoLR9FpbnwRmTa69yLHF3QfcaxbWT7gWdwws5k4dpmJvqpEuMWwnj

  bip32.BIP32 nodeFromSeed = bip32.BIP32
      .fromSeed(HEX.decode("000102030405060708090a0b0c0d0e0f") as Uint8List);
  print(nodeFromSeed.toBase58());
  // => xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi

  bip32.BIP32 nodeFromPub = bip32.BIP32.fromBase58(
      "xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8");
  print(nodeFromPub.toBase58());
  // => xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8

  var message = HEX.decode(
          "0202020202020202020202020202020202020202020202020202020202020202")
      as Uint8List;
  var signature = nodeFromSeed.sign(message);
  print(signature);
  // => [63, 219, 20, 114, 95, 184, 192, 55, 216, 206, 126, 121, 17, 71, 64, 70, 163, 82, 247, 73, 243, 95, 30, 137, 177, 155, 100, 225, 177, 203, 217, 147, 122, 64, 208, 129, 54, 133, 113, 41, 216, 160, 191, 15, 136, 98, 235, 25, 219, 178, 70, 222, 127, 151, 135, 242, 25, 192, 161, 187, 187, 84, 81, 215]

  print(HEX.encode(signature));
  // => 3fdb14725fb8c037d8ce7e7911474046a352f749f35f1e89b19b64e1b1cbd9937a40d08136857129d8a0bf0f8862eb19dbb246de7f9787f219c0a1bbbb5451d7

  print(nodeFromSeed.verify(message, signature));
  // => true
}
